File: 2023/rust/src/bin/d08p2.rs
use std::{collections::HashMap, fs};

type Map<'a> = HashMap<&'a str, Node<'a>>;

#[derive(Debug)]
enum Direction {
    Left,
    Right,
}
impl From<char> for Direction {
    fn from(value: char) -> Self {
        match value {
            'L' => Direction::Left,
            'R' => Direction::Right,
            _ => panic!("Unknown direction"),
        }
    }
}

#[derive(Debug)]
struct Node<'a> {
    name: &'a str,
    left: &'a str,
    right: &'a str,
}
impl<'a> Node<'a> {
    fn from_str(s: &'a str) -> Result<Self, ()> {
        let (name, (left, right)) = s
            .split_once(" = ")
            .and_then(|s| {
                Some((
                    s.0,
                    s.1.strip_prefix('(')
                        .and_then(|s| s.strip_suffix(')'))
                        .and_then(|s| s.split_once(", "))?,
                ))
            })
            .ok_or(())?;

        Ok(Node { name, left, right })
    }
}

fn main() {
    println!("Day 8 - Part 2");

    let str = fs::read_to_string("../_input/day08.txt").unwrap();
    let lines: Vec<&str> = str.lines().collect();

    let total = process(lines);
    println!("Total: {}", total);
}

fn process(lines: Vec<&str>) -> u64 {
    let instructions: Vec<Direction> = lines[0].chars().map(|i| i.into()).collect();
    let (mut starting, map) = process_map(&lines[2..lines.len()]);

    let steps: Vec<u64> = starting
        .iter_mut()
        .map(|current_node| {
            let mut current_dir = 0;
            let mut hit = false;
            let mut counter = 0;

            while !hit {
                let node = map.get(current_node).unwrap();
                match instructions[current_dir] {
                    Direction::Left => *current_node = node.left,
                    Direction::Right => *current_node = node.right,
                }

                counter += 1;
                if current_node.ends_with('Z') {
                    hit = true;
                }

                current_dir += 1;
                if current_dir >= instructions.len() {
                    current_dir = 0;
                }
            }

            counter
        })
        .collect();

    steps.into_iter().reduce(lcm).unwrap()
}

fn lcm(a: u64, b: u64) -> u64 {
    (a * b) / gcd(a, b)
}

fn gcd(a: u64, b: u64) -> u64 {
    if b == 0 {
        a
    } else {
        gcd(b, a % b)
    }
}

fn process_map<'a>(lines: &'a [&'a str]) -> (Vec<&'a str>, Map<'a>) {
    let mut map = Map::new();
    let mut starting = Vec::new();

    for s in lines {
        let node = Node::from_str(s).unwrap();
        if node.name.ends_with('A') {
            starting.push(node.name);
        }
        map.insert(node.name, node);
    }

    (starting, map)
}

#[cfg(test)]
mod test {
    use crate::process;

    #[test]
    fn test() {
        assert_eq!(
            process(
                r"RL

AAA = (BBB, CCC)
BBB = (DDD, EEE)
CCC = (ZZZ, GGG)
DDD = (DDD, DDD)
EEE = (EEE, EEE)
GGG = (GGG, GGG)
ZZZ = (ZZZ, ZZZ)"
                    .split('\n')
                    .collect()
            ),
            2
        );

        assert_eq!(
            process(
                r"LLR

AAA = (BBB, BBB)
BBB = (AAA, ZZZ)
ZZZ = (ZZZ, ZZZ)"
                    .split('\n')
                    .collect()
            ),
            6
        );
    }
}

File: 2023/swift/d08p2.swift
import Foundation

enum Direction: Character {
    case left = "L"
    case right = "R"
}

struct Node {
    let name: String
    let left: String
    let right: String

    static func fromString(_ string: String) -> Self {
        fromString(Substring(string))
    }

    static func fromString(_ string: Substring) -> Self {
        let s1 = string.split(separator: " = ", maxSplits: 1)
        let s2 = s1[1].trimmingCharacters(in: CharacterSet(charactersIn: "()")).split(
            separator: ", ", maxSplits: 1)
        return Self(name: String(s1[0]), left: String(s2[0]), right: String(s2[1]))
    }
}

func main() {
    print("Day 8 - Part 2")

    let raw_data = try! NSString(
        contentsOfFile: "../_input/day08.txt", encoding: String.Encoding.utf8.rawValue)
    let data = raw_data as String

    let total = process(String(data))
    print("Total: \(total)")
}

func process(_ string: String) -> Int {
    let input = string.trimmingCharacters(in: CharacterSet.whitespacesAndNewlines).split(
        separator: "\n")
    let instructions = input[0].map({ Direction(rawValue: $0)! })
    let (starting, map) = processMap(input[1..<input.endIndex])

    let steps = starting.map {
        var currentNode = $0
        var currentDir = 0
        var hit = false
        var counter = 0
        while !hit {
            let node = map[currentNode]!
            switch instructions[currentDir] {
            case .left:
                currentNode = node.left
            case .right:
                currentNode = node.right
            }

            counter += 1
            if currentNode.hasSuffix("Z") {
                hit = true
            }

            currentDir += 1
            if currentDir >= instructions.count {
                currentDir = 0
            }
        }

        return counter
    }

    return steps.reduce(steps[0]) { lcm($0, $1) }
}

func lcm(_ a: Int, _ b: Int) -> Int {
    return (a * b) / gcd(a, b)
}

func gcd(_ a: Int, _ b: Int) -> Int {
    return b == 0 ? a : gcd(b, a % b)
}

func processMap(_ string: ArraySlice<Substring>) -> ([String], [String: Node]) {
    var map: [String: Node] = [:]
    var starting: [String] = []

    for s in string {
        let node = Node.fromString(s)
        if node.name.hasSuffix("A") { starting.append(node.name) }
        map[node.name] = node
    }

    return (starting, map)
}

main()
